src/core/init/stream_events_emitter.ts

```diff
@@ -0,0 +1,160 @@
+ * limitations under the License.
+ */
+
+import {
+  EMPTY,
+  interval,
+  Observable,
+  of as observableOf,
+} from "rxjs";
+import {
+  filter,
+  finalize,
+  mergeMap,
+} from "rxjs/operators";
+import config from "../../config";
+import Manifest from "../../manifest";
+import arrayFindIndex from "../../utils/array_find_index";
+
+const { STREAM_EVENT_EMITTER_POLL_INTERVAL } = config;
+
+export interface IScheduleEvent {
+  start: number;
+  end?: number;
+  id?: string;
+  element: Node | Element;
+}
+
+interface IGoingInStreamEvent {
+  type: "going-in";
+  value: IScheduleEvent;
+}
+
+interface IGoingOutStreamEvent {
+  type: "going-out";
+  value: IScheduleEvent;
+}
+
+export type IEmittedStreamEvent = IGoingInStreamEvent |
+                                  IGoingOutStreamEvent;
+
+/**
+ * Compare 2 events.
+ * @param {Object} evt1
+ * @param {Object} evt2
+ * @returns {Boolean}
+ */
+function areSameStreamEvents(evt1: IScheduleEvent,
+                             evt2: IScheduleEvent): boolean {
+  return evt1.id === evt2.id &&
+         evt1.start === evt2.start &&
+         evt1.end === evt2.end;
+}
+
+/**
+ * Get events from manifest and emit each time an event has to be emitted
+ * @param {Object} manifest
+ * @param {HTMLMediaElement} mediaElement
+ */
+function streamEventsEmitter(manifest: Manifest,
+                             mediaElement: HTMLMediaElement
+): Observable<IEmittedStreamEvent> {
+  let scheduledEvents: IScheduleEvent[] = [];
+  const currentEventsIn: IScheduleEvent[] = [];
+
+  /**
+   * Refresh local scheduled events list
+   */
+  function updateScheduledEvents(): void {
+    scheduledEvents = [];
+    const { periods } = manifest;
+    for (let i = 0; i < periods.length; i++) {
+      const period = periods[i];
+      const { streamEvents } = period;
+      if (streamEvents !== undefined) {
+        streamEvents.forEach((event) => {
+          const start = period.start + (event.presentationTime ?? 0);
+          const end = event.duration !== undefined ?
+            start + event.duration :
+            undefined;
+          scheduledEvents.push({ start,
+                                 end,
+                                 id: event.id,
+                                 element: event.element });
+        });
+      }
+    }
+    for (let j = 0; j < currentEventsIn.length; j++) {
+      const evt = currentEventsIn[j];
+      if (!scheduledEvents.some((s) => areSameStreamEvents(s, evt))) {
+        currentEventsIn.splice(j, 1);
+        j--;
+      }
+    }
+  }
+
+  updateScheduledEvents();
+  manifest.addEventListener("manifestUpdate", updateScheduledEvents);
+
+  return interval(STREAM_EVENT_EMITTER_POLL_INTERVAL)
+    .pipe(
+      filter(() => scheduledEvents.length !== 0),
+      mergeMap(() => {
+        const { currentTime } = mediaElement;
+        const eventsToSend: IEmittedStreamEvent[] = [];
+        for (let i = 0; i < scheduledEvents.length; i++) {
```

Comment:
Because currentTime may change due to a seeking, we may go out multiple events at one time, thus we can't abort the loop, because we may miss some events to emit.

Reply:
That's true.

Maybe we could make it so we first iterate on every chronological event and abort and then iterate on missed active events.